enhancements/api-review/add-unqualified-search-registries.md

```diff
@@ -0,0 +1,211 @@
+
+### Non-Goals
+
+## Proposal
+
+The Image API is extended by adding an optional `UnqualifiedSearchRegistries` field with type `[]string` to `RegistrySources`:
+
+```go
+// RegistrySources holds cluster-wide information about how to handle the registries config.
+type RegistrySources struct {
+    // ...
+    
+	// unqualifiedSearchRegistries are registries that will be searched when pulling images using short names.
+	// +optional
+	UnqualifiedSearchRegistries []string `json:"unqualifiedSearchRegistries,omitempty"`
+}
+```
+
+The containerRuntimeConfig controller in the MCO already watches the cluster wide images.config.openshift.io CR for the allowed, blocked, and insecure registries. It will now watch for `unqualified-search-registries` and update `/etc/containers/registries.conf` accordingly.
+
+An example images.config.openshift.io CR will look like:
+```yaml
+apiVersion: config.openshift.io/v1
+kind: Image
+metadata:
+  name: cluster
+spec:
+  registrySources:
+    unqualifiedSearchRegistries:
+    - "reg1.io"
+    - "reg2.io"
+    - "reg3.io"
+```
+
+### User Stories
+
+#### As a user, I would like to use image short names when running my workloads
+
+The user can set the `unqualified-search-registries` with a list of registries to check when pulling an image using short names.
+The user can run `oc edit images.config.openshift.io cluster` and add `UnqualifiedSearchRegistries` under `RegistrySources`. Once this is done, the containerRuntimeConfig controller will roll out the changes to the nodes.
+
+#### As a user, I would like to use multiple internal registries to pull my images
+
+The user can use multiple internal registries to pull images with short names without having to change the image spec every time that the registries' DNS changes. This can be done by configuring the list of `unqualified-search-registries` to reflect the changes in the internal registries names by running `oc edit images.config.openshift.io cluster` and add `UnqualifiedSearchRegistries` under `RegistrySources`. Once this is done, the containerRuntimeConfig controller will roll out the changes to the nodes.
+
+### Implementation Details/Notes/Constraints
+
+Implementing this enhancement requires changes in:
+- openshift/api
+- openshift/machine-config-operator
+
+This is what the `/etc/containers/registries.conf` file currently looks like on the nodes:
+
+```
+unqualified-search-registries = ['registry.access.redhat.com', 'docker.io']
+
+[[registry]]
+...
+```
+
+This is an example of the cluster wide images.config.openshift.io:
+
+```
+apiVersion: config.openshift.io/v1
+kind: Image
+metadata:
+  name: cluster
+spec:
+  registrySources:
+    unqualifiedSearchRegistries:
+    - "reg1.io"
+    - "reg2.io"
+    - "reg3.io"
+```
+
+The above Image CR will result in the `/etc/containers/registries.conf` file on each node to look like:
+
+```
+unqualified-search-registries = ['registry.access.redhat.com', 'docker.io', 'reg1.io', 'reg2.io', 'reg3.io']
```

Comment:
We need to consider cases where users may want to remove these registries or modify their priority in the search order. Maybe, the behavior should be to overwrite the entire list with a big warning in the docs.

Reply:
We can overwrite the entire list by dropping a file in `/etc/containers/registries.conf.d`, but it won't fall back to search the default list in `/etc/containers/registries.conf` if the image is not found in one of the registries listed in the drop-in file.

Reply:
Fixed